vcenter_operator/vault.py

```diff
@@ -0,0 +1,262 @@
+
+        if resp.status_code >= 500:
+            raise VaultUnavailableError()
+
+        resp.raise_for_status()
+
+        auth_data = resp.json().get("auth", {})
+        return auth_data.get("client_token"), auth_data.get("lease_duration")
+
+    def get_secret(self, path):
+        """Get the secret from vault"""
+        self._check_parameters_set()
+
+        headers = self._get_headers()
+        resp = requests.get(f"{self.vault_url}/v1/{self.mount_point}/data/{path}", headers=headers)
+
+        if resp.status_code >= 500:
+            raise VaultUnavailableError()
+
+        resp.raise_for_status()
+
+        return resp.json().get("data", {}).get("data")
+
+    def get_metadata(self, path):
+        """Get the metadata of the secret"""
+        self._check_parameters_set()
+
+        headers = self._get_headers()
+        resp = requests.get(f"{self.vault_url}/v1/{self.mount_point}/metadata/{path}", headers=headers)
+
+        if resp.status_code >= 500:
+            raise VaultUnavailableError()
+
+        if resp.status_code == 404:
+            LOG.warning(f"Could not find the secret under path {path} in vault")
+            return None
+
+        resp.raise_for_status()
+
+        return resp.json()
+
+    def create_service_user(self, username_template, path, service, last_version=None):
+        """Create the service-user"""
+        self._check_parameters_set()
+
+        # Initial username starting with 0001 (version 1)
+        # When rotating passwords, username is incremented by 1 for parallel use
+        username = (
+            username_template + "0001"
+            if last_version is None
+            else username_template + str(int(last_version) + 1).zfill(4)
+        )
+
+        password = self.gen_password()
+
+        if self.dry_run:
+            LOG.debug("Dry-run: Would have created service-user")
+            return "1", username, password
+
+        version = self.store_service_user_credentials(username, password, path, service)
+
+        return version, username, password
+
+    def gen_password(self):
+        """
+        Generate a password with the given constraints.
+        - length: total length of the password
+        - digits: number of digits
+        - symbols: number of symbols
+        """
+        if PASSWORD_LENGTH < PASSWORD_DIGITS + PASSWORD_SYMBOLS:
+            raise ValueError("Password length too short for the required digits and symbols.")
+
+        password_chars = [
+            random.choice(string.digits) for _ in range(PASSWORD_DIGITS)
+        ] + [
+            random.choice(string.punctuation) for _ in range(PASSWORD_SYMBOLS)
+        ]
+
+        remaining_length = PASSWORD_LENGTH - PASSWORD_DIGITS - PASSWORD_SYMBOLS
+        password_chars += [
+            random.choice(string.ascii_letters) for _ in range(remaining_length)
+        ]
+
+        random.shuffle(password_chars)
+        return ''.join(password_chars)
```

Comment:
This needs to be `secrets` module instead of `random` for security purposes - which password generation is for me. I'd prefer if we can get the password created by Vault.